### PR TITLE
impl(GCS+gRPC): add mock for `AsyncConnection`

### PR DESCRIPTION
**Pull request by coryan**
Also includes some examples of how to use the mock.  These double up as tests.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12806)

### CODE REVIEW

File: BUILD.bazel [hunk 1]
```diff
@@ -160,6 +160,16 @@ cc_library(
     ],
 )
 
+cc_library(
+    name = "experimental-storage_grpc_mocks",
+    testonly = True,
+    hdrs = ["//google/cloud/storage:grpc_mocks_hdrs"],
+    tags = ["manual"],
```
Comment (devbww) [inline BUILD.bazel:167]: Is this applicable to a `cc_library`?

File: google/cloud/storage/examples/CMakeLists.txt [hunk 2]
```diff
@@ -108,3 +106,48 @@ foreach (fname ${storage_examples_production})
         ${target} PROPERTIES LABELS
                              "integration-test;integration-test-production")
 endforeach ()
+
+# These lists are always defined and exported to Bazel. They are only *used* if
+# experimental-storage_grpc is defined.
+set(storage_grpc_examples # cmake-format: sort
+                          storage_async_samples.cc storage_grpc_samples.cc)
+# The mocking examples take the form of a unit test
```
Comment (devbww) [inline google/cloud/storage/examples/CMakeLists.txt:114]: Nit: `s/$/./`

File: google/cloud/storage/examples/storage_async_mock_samples.cc [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  // To simplify the test we use .get() to "block" until the gc::future is
+  // ready, and then use `.value()` to get the `StatusOr<>` values or an
+  // exception.
+  gcs_ex::AsyncReader reader;
+  gcs_ex::AsyncToken token;
+  std::tie(reader, token) =
+      client.ReadObject("test-bucket", "test-object").get().value();
+
+  gcs_ex::ReadPayload payload;
+  gcs_ex::AsyncToken t;  // use a temporary to make clang-tidy happy.
```
Comment (devbww) [inline google/cloud/storage/examples/storage_async_mock_samples.cc:85]: "non-temporary" perhaps?

**Review by devbww — APPROVED** [review]

File: BUILD.bazel (hunk 1 above)
Comment (coryan) [reply BUILD.bazel]: Removed.

File: google/cloud/storage/examples/CMakeLists.txt (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/examples/CMakeLists.txt]: Fixed.

**Pull request merged**